src/Arcus.WebApi.Logging.Core/Correlation/HttpCorrelationService.cs

```diff
@@ -0,0 +1,151 @@
+            Guard.NotNull(options.Value, nameof(options.Value), "Requires a value in the set of options to configure the correlation process");
+            
+            _httpContextAccessor = httpContextAccessor;
+            _options = options.Value;
+            _logger = logger;
+        }
+
+        /// <summary>
+        /// Correlate the current HTTP request according to the previously configured correlation options; returning an <paramref name="errorMessage"/> when the correlation failed.
+        /// </summary>
+        /// <param name="errorMessage">The failure message that describes why the correlation of the HTTP request wasn't successful.</param>
+        /// <returns>
+        ///     [true] when the HTTP request was successfully correlated and the HTTP response was altered accordingly;
+        ///     [false] there was a problem with the correlation, describing the failure in the <paramref name="errorMessage"/>.
+        /// </returns>
+        public bool TryHttpCorrelate(out string errorMessage)
+        {
+            HttpContext httpContext = _httpContextAccessor.HttpContext;
+
+            Guard.NotNull(httpContext, nameof(httpContext), "Requires a HTTP context from the HTTP context accessor to start correlating the HTTP request");
+            Guard.For<ArgumentException>(() => httpContext.Response is null, "Requires a 'Response'");
+            Guard.For<ArgumentException>(() => httpContext.Response.Headers is null, "Requires a 'Response' object with headers");
+
+            if (httpContext.Request.Headers.TryGetValue(_options.Transaction.HeaderName, out StringValues transactionIds))
+            {
+                if (!_options.Transaction.AllowInRequest)
+                {
+                    _logger.LogError("No correlation request header '{HeaderName}' for transaction ID was allowed in request", _options.Transaction.HeaderName);
+
+                    errorMessage = $"No correlation transaction ID request header '{_options.Transaction.HeaderName}' was allowed in the request";
+                    return false;
+                }
+
+                _logger.LogTrace("Correlation request header '{HeaderName}' found with transaction ID '{TransactionId}'", _options.Transaction.HeaderName, transactionIds);
+            }
+
+            string operationId = DetermineOperationId(httpContext);
+            string transactionId = DetermineTransactionId(transactionIds);
+            var correlation = new CorrelationInfo(operationId, transactionId);
+            httpContext.Features.Set(correlation);
+
+            AddCorrelationResponseHeaders(httpContext, operationId, transactionId);
+
+            errorMessage = null;
+            return true;
+        }
+
+        private string DetermineOperationId(HttpContext httpContext)
+        {
+            if (String.IsNullOrWhiteSpace(httpContext.TraceIdentifier))
```

Comment:
And this is the fun part of Azure Functions, they do not use this approach (unless this changed); see https://github.com/Azure/azure-functions-host/issues/5014

That's why we are using this for Azure Functions:
```csharp
private string GetOperationIdFromRequestHeaders(ExecutionContext executionContext)
{
    return Activity.Current?.Id ?? executionContext.InvocationId.ToString();
}
```

Reply:
So we can use this for sure, but you'll need to use different operation id providers depending on the runtime

Reply:
Hmm, I think I got a valid reponse back with the `.TraceIdentifier` filled out. 🤔 .

Reply:
Do you have the function in a repo somewhere?

Reply:
No, but I'll look if I can maybe create an integration test for this; with an embedded Azure Function.

Reply:
No, not that. Hmm, am thinking what would be the best approach to test this.

Reply:
I would add an Azure Function to the solution which we start in the pipeline and send request to, I would not start if from our code.

Reply:
You can use the same approach locally.